src/rgw/driver/rados/rgw_pubsub.h

```diff
@@ -506,7 +504,7 @@ struct rgw_pubsub_bucket_topics {
 };
 WRITE_CLASS_ENCODER(rgw_pubsub_bucket_topics)
 
-struct rgw_pubsub_topics {
+struct rgw_pubsub_topics : public rgw::sal::NotificationConfig::Result {
```

Comment:
do you think these topic structs are sufficiently rados-specific that they need to be hidden under this `Result` interface? maybe these should be above zipper?

or is this abstraction just to support both structs `rgw_pubsub_topics` and `rgw_pubsub_bucket_topics`? if so, then how does the sal interface prevent a caller from providing the wrong one? maybe we need two interfaces, `NotificationConfig` and `BucketNotificationConfig`? or maybe we can avoid having to allocate a `NotificationConfig` interface by exposing the read/write/delete functions directly under `Driver` and `Bucket`:
```c++
class Driver {
 public:
  virtual int read_notification_config(std::string_view tenant, rgw_pubsub_topics& topics,
	      RGWObjVersionTracker* objv_tracker, optional_yield y, const DoutPrefixProvider *dpp) = 0;
  virtual int write_notification_config(std::string_view tenant, const rgw_pubsub_topics& topics,
	      RGWObjVersionTracker* obj_tracker, optional_yield y, const DoutPrefixProvider *dpp) = 0;
  virtual int remove_notification_config(std::string_view tenant, RGWObjVersionTracker* objv_tracker,
	      optional_yield y, const DoutPrefixProvider *dpp) = 0;
```

Reply:
> do you think these topic structs are sufficiently rados-specific that they need to be hidden under this `Result` interface? maybe these should be above zipper?
> 
> or is this abstraction just to support both structs `rgw_pubsub_topics` and `rgw_pubsub_bucket_topics`?

i think that this was the intent i nthe code i took from [here](https://github.com/ceph/ceph/pull/48814)

> if so, then how does the sal interface prevent a caller from providing the wrong one? maybe we need two interfaces, `NotificationConfig` and `BucketNotificationConfig`? 

in the original implementation (before zipper) the functions were templatized with the actual structs.
since there are only 2 versions of the functions and they are identical except for the structs, i think it makes sense to just templatize them.
I would need to move `rgw_pubsub.h/cc`` above the zipper line, but this should be ok now

> or maybe we can avoid having to allocate a `NotificationConfig` interface by exposing the read/write/delete functions directly under `Driver` and `Bucket`:
> 
> ```c++
> class Driver {
>  public:
>   virtual int read_notification_config(std::string_view tenant, rgw_pubsub_topics& topics,
> 	      RGWObjVersionTracker* objv_tracker, optional_yield y, const DoutPrefixProvider *dpp) = 0;
>   virtual int write_notification_config(std::string_view tenant, const rgw_pubsub_topics& topics,
> 	      RGWObjVersionTracker* obj_tracker, optional_yield y, const DoutPrefixProvider *dpp) = 0;
>   virtual int remove_notification_config(std::string_view tenant, RGWObjVersionTracker* objv_tracker,
> 	      optional_yield y, const DoutPrefixProvider *dpp) = 0;
> ```

makes sense. @dang any reason to separate into its own class?

Reply:
* removed the notificationConfig class hierarchy
* added the Store/Bucket APIs as casey suggested
* just duplicated the code between the read/write/remove functions for the 2 different structs - if this is an issue i can templatize them with the struct type